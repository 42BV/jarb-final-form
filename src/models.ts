export type FieldType =
  | 'enum'
  | 'color'
  | 'datetime-local'
  | 'datetime'
  | 'month'
  | 'week'
  | 'date'
  | 'time'
  | 'email'
  | 'tel'
  | 'number'
  | 'url'
  | 'password'
  | 'file'
  | 'image'
  | 'boolean'
  | 'text';

export type FieldConstraints = {
  javaType: string;
  types?: FieldType[] | null;
  required?: boolean | null;
  minimumLength?: number | null;
  maximumLength?: number | null;
  fractionLength?: number | null;
  radix?: number | null;
  pattern?: string | null;
  min?: number | null;
  max?: number | null;
  name: string;
};

export type ConstraintModel = Record<string, FieldConstraints>;

/**
 * The constraints should have the following signature:
 *
 * {
 *   "SuperHero": {
 *     "name": {
 *       "javaType": "java.lang.String",
 *       "types": ["text"],
 *       "required": true,
 *       "minimumLength": null,
 *       "maximumLength": 50,
 *       "fractionLength": null,
 *       "radix": null,
 *       "pattern": null,
 *       "min": null,
 *       "max": null,
 *       "name": "name"
 *     },
 *     "email": {
 *       "javaType": "java.lang.String",
 *       "types": ["email", "text"],
 *       "required": true,
 *       "minimumLength": null,
 *       "maximumLength": 255,
 *       "fractionLength": null,
 *       "radix": null,
 *       "pattern": null,
 *       "min": null,
 *       "max": null,
 *       "name": "email"
 *     }
 * }
 *
 * The keys represent the name of the class, in the above case 'SuperHero', each class
 * has fields such as the 'name', and 'email', these are described in an object of the
 * same name. These 'validator' objects look like this:
 *
 * {
 *   "javaType": string,          // The Java class name of this validator
 *   "types":Array<string>,       // The type that closest represents this validator
 *   "required":true,             // Wether or not the validator is required.
 *   "minimumLength":int,         // Minimum length of the input string.
 *   "maximumLength":int,         // Maximum length of the input string.
 *   "fractionLength":int,        // The number of numbers after the dot if input is a number.
 *   "radix": int,                // Radix for the when type is number: @See http://en.wikipedia.org/wiki/Radix. Is not used.
 *   "pattern": string,           // The regex in Java form the input must be valid for. Is not used.
 *   "min": int,                  // The maximum int value, is not used.
 *   "max": int,                  // The minimum int value, is not used.
 *   "name": string               // The name of the property this validator represents.
 * }
 *
 */
export type Constraints = {
  [key: string]: ConstraintModel;
};
