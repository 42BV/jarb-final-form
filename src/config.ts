export interface Config {
  // The URL which will provide the constraints over a GET request.
  constraintsUrl: string;
}

let config: Config | null = null;

/**
 * Configures the Constraint libary.
 *
 * @param {Config} c The new configuration
 */
export function configureConstraint(c: Config): void {
  config = c;
}

/**
 * Either returns the a Config or throws an error when the
 * config is not yet initialized.
 *
 * @returns The Config
 */
export function getConfig(): Config {
  if (config === null) {
    throw new Error('The constraint service is not initialized.');
  } else {
    return config;
  }
}
