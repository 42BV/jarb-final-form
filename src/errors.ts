export type RequiredError = {
  type: 'ERROR_REQUIRED';
  label: string;
  value: unknown;
  reasons: {
    required: 'required';
  };
};

export type MinimumLengthError = {
  type: 'ERROR_MINIMUM_LENGTH';
  label: string;
  value: unknown;
  reasons: {
    minimumLength: number;
  };
};

export type MaximumLengthError = {
  type: 'ERROR_MAXIMUM_LENGTH';
  label: string;
  value: unknown;
  reasons: {
    maximumLength: number;
  };
};

export type MinValueError = {
  type: 'ERROR_MIN_VALUE';
  label: string;
  value: unknown;
  reasons: {
    minValue: number;
  };
};

export type MaxValueError = {
  type: 'ERROR_MAX_VALUE';
  label: string;
  value: unknown;
  reasons: {
    maxValue: number;
  };
};

export type NumberError = {
  type: 'ERROR_NUMBER';
  label: string;
  value: unknown;
  reasons: {
    regex: RegExp;
  };
};

export type NumberFractionError = {
  type: 'ERROR_NUMBER_FRACTION';
  label: string;
  value: unknown;
  reasons: {
    regex: RegExp;
    fractionLength: number;
  };
};

export type ValidationError =
  | RequiredError
  | MinimumLengthError
  | MaximumLengthError
  | MinValueError
  | MaxValueError
  | NumberError
  | NumberFractionError;
