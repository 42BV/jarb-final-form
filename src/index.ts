export { configureConstraint } from './config';
export { JarbField, JarbFieldProps, JarbProps } from './JarbField';
export { loadConstraints, getConstraints, setConstraints } from './constraints';
export { FieldType, Constraints } from './models';
export {
  RequiredError,
  MinimumLengthError,
  MaximumLengthError,
  MinValueError,
  MaxValueError,
  NumberError,
  NumberFractionError,
  ValidationError
} from './errors';
