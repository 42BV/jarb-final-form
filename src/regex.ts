/**
 * Returns a regex that checks if the it is a valid number
 * which can have fractions. Which are the numbers behind
 * the decimal. So if the fractionLength is 5 you accept:
 * #.#####, which means 5 numbers after the decimals.
 *
 * The number can be negative or positive.
 *
 * @param  {number} fractionLength The length of the fraction which is considered valid.
 * @return {regex}                 A regex which checks for fraction numbers.
 */
export function fractionNumberRegex(fractionLength: number): RegExp {
  return new RegExp('^-?\\d+(\\.\\d{1,' + fractionLength + '})?$');
}

/**
 * A regex which checks for a positive or negative number
 * without fractions.
 */
export const numberRegex: RegExp = /^-?\d+$/;
