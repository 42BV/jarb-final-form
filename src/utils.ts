import {
  FieldType,
  Constraints,
  ConstraintModel,
  FieldConstraints
} from './models';

// List of <input> types sorted on most specific first.
const inputTypes: FieldType[] = [
  'color',
  'datetime-local',
  'datetime',
  'month',
  'week',
  'date',
  'time',
  'email',
  'tel',
  'number',
  'url',
  'password',
  'file',
  'image',
  'text'
];

/**
 * Finds the most specific <input> type for the types parameter. For example if
 * types is ['email', 'text'] the function returns 'email' because 'email'
 * is the most specific input type. If nothing is found returns 'text'.
 *
 * @param  {Array<string>} The types you want the closest type for.
 * @return {FieldType} The closest <input> type, based on the types parameter.
 */
export function mostSpecificInputTypeFor(types: FieldType[]): FieldType {
  // Default to the last inputType which should be 'text'.
  let index = inputTypes.length - 1;

  for (let i = 0; i < types.length; i += 1) {
    const type = types[i];
    for (let j = 0; j < inputTypes.length; j += 1) {
      const inputType = inputTypes[j];

      //console.log(`${type} === ${inputType}`);
      if (type === inputType) {
        index = Math.min(index, j);
        break;
      }
    }
  }

  return inputTypes[index];
}

/**
 * Finds the FieldConstraints rules for a specific validator in the
 * Constraints object.
 *
 * If no constraints can be found for a validator the boolean false
 * is returned.
 *
 * @param  {validator} 'validator' is a string with the format: 'Class.field' for example: 'User.age'
 * @param  {Constraints} The constraints to find the validator in.
 * @throws {error} When the validator doesn't match the format 'className.fieldName'.
 * @returns {FieldConstraints | false} The constraints for the specific field
 */
export function getFieldConstraintsFor(
  validator: string,
  constraints: Constraints
): FieldConstraints | false {
  const [className] = validator.split('.', 1);
  const propertyName = validator.substring(className.length + 1);

  const classConstraints: ConstraintModel = constraints[className];

  if (classConstraints !== undefined) {
    const fieldConstraints = classConstraints[propertyName];

    return fieldConstraints !== undefined ? fieldConstraints : false;
  } else {
    return false;
  }
}
