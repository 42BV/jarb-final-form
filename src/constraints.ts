import { getConfig } from './config';
import { Constraints } from './models';
import { get } from '@42.nl/spring-connect';

let constraints: Constraints | undefined = undefined;

/**
 * Loads the constraints from the back-end.
 *
 * The URL it will send the request to is defined by the 'constraintsUrl'
 * from the Config object. The HTTP method it uses is 'get'.
 *
 * The entire response will be written to the `constraints` variable.
 * Whatever the JSON response is will be the constraints.
 *
 *  An example response:
 *
 * ```JSON
 * {
 *   "SuperHero": {
 *     "name": {
 *       "javaType": "java.lang.String",
 *       "types": ["text"],
 *       "required": true,
 *       "minimumLength": null,
 *       "maximumLength": 50,
 *       "fractionLength": null,
 *       "radix": null,
 *       "pattern": null,
 *       "min": null,
 *       "max": null,
 *       "name": "name"
 *     },
 *     "email": {
 *       "javaType": "java.lang.String",
 *       "types": ["email", "text"],
 *       "required": true,
 *       "minimumLength": null,
 *       "maximumLength": 255,
 *       "fractionLength": null,
 *       "radix": null,
 *       "pattern": null,
 *       "min": null,
 *       "max": null,
 *       "name": "email"
 *     }
 *   }
 * }
 * ```
 *
 * @returns {Promise}
 */
export async function loadConstraints(): Promise<void> {
  const { constraintsUrl } = getConfig();

  constraints = await get<Constraints>(constraintsUrl);
}

/**
 * Sets the constraints.
 *
 * @param newConstraints The new constraints
 */
export function setConstraints(newConstraints: Constraints | undefined): void {
  constraints = newConstraints;
}

/**
 * Get the current constraints.
 *
 * @returns The current constraints
 */
export function getConstraints(): Constraints | undefined {
  return constraints;
}
