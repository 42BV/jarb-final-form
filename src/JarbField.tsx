import React, { Component } from 'react';
import { getConstraints } from './constraints';
import { FieldValidator, FieldState } from 'final-form';
import { Field, FieldProps, FieldRenderProps } from 'react-final-form';

import { getFieldConstraintsFor, mostSpecificInputTypeFor } from './utils';
import { FieldType } from './models';
import * as Validators from './validators';

export interface JarbProps {
  validator: string;
  label: string;
}
export interface JarbFieldProps<FieldValue, T extends HTMLElement>
  extends FieldProps<FieldValue, FieldRenderProps<FieldValue>, T> {
  jarb: JarbProps;

  /**
   * An array of custom validators to run whenever the field changes.
   *
   * @type {FieldValidator<FieldValue>[]}
   * @memberof JarbFieldProps
   */
  validators?: FieldValidator<FieldValue>[];

  /**
   * An array of custom async validators to run whenever the field changes
   * and the synchronous validators have declared the field valid.
   *
   * @type {FieldValidator<FieldValue>[]}
   * @memberof JarbFieldProps
   */
  asyncValidators?: FieldValidator<FieldValue>[];

  /**
   * The time after which the async validators are ran to prevent
   * to many validators from running at once.
   *
   * Defaults to 200 milliseconds.
   *
   * @type {number}
   * @memberof JarbFieldProps
   */
  asyncValidatorsDebounce?: number;
}

/**
 * JarbField wraps final-form's Field, and adds the auto validation
 * from the constraints. In fact it is a very thin wrapper around
 * Field.
 *
 * It only demands one extra property called `jarb` which is used
 * to to configure the Field. The `jarb` object needs two keys:
 * the `validator` and the `label`. 
 *
 * The `validator` follows the following format: `{Entity}.{Property}`. 
 * For example if the validator property is 'SuperHero.name' this means that
 * the Field will apply the constraints for the 'name' property of
 * the `SuperHero` entity.
 * 
 * The `label` is used to inform you which field was wrong, when errors occur.
 * You will receive the 'label' when an error occurs to create a nice
 * error message.
 * 
 * It is also possible to add custom field validators and async validators
 * via the `validators` and `asyncValidators` props. The `asyncValidators`
 * are only ran when the all synchronous `validators` have declared the
 * field valid.
 * 
 * See the documentation for some examples on how to create custom
 * validators.
 *
 * @example
 * ```JavaScript
 * <JarbField 
 *   name="Name" 
 *   jarb={{ validator: 'SuperHero.name', label: "Name" }} 
 *   component="input" 
 *   type="text"
 * />
 * ```
 *
 * @export
 * @param {Props.jarb} object Object containing the 'label' and 'validator'.

 * @returns
 */
export class JarbField<FieldValue, T extends HTMLElement> extends Component<
  JarbFieldProps<FieldValue, T>
> {
  private enhancedValidate: FieldValidator<FieldValue> | null = null;

  // Stores a resolver which is used to debounce async validations
  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
  private debounceResolver = (
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    value?: boolean | PromiseLike<boolean> | undefined
  ) => {}; // eslint-disable-line @typescript-eslint/no-empty-function

  // Stores a ID for each async validation train.
  private validationId = Math.random();

  public getEnhancedValidate(): FieldValidator<FieldValue> | null {
    if (this.enhancedValidate !== null) {
      return this.enhancedValidate;
    }

    const {
      jarb,
      validators,
      asyncValidators,
      asyncValidatorsDebounce = 200
    } = this.props;
    const { label, validator } = jarb;

    const validatorFunctions =
      Array.isArray(validators) && validators ? [...validators] : [];

    const asyncValidatorFunctions =
      Array.isArray(asyncValidators) && asyncValidators
        ? [...asyncValidators]
        : [];

    const constraints = getConstraints();

    if (constraints !== undefined) {
      const fieldConstraints = getFieldConstraintsFor(validator, constraints);

      if (fieldConstraints !== false) {
        const field: FieldType = mostSpecificInputTypeFor(
          fieldConstraints.types
        );

        if (fieldConstraints.required) {
          if (field === 'boolean') {
            const requiredValidator = Validators.makeBooleanRequired(label);
            validatorFunctions.push(requiredValidator);
          } else {
            const requiredValidator = Validators.makeRequired(label);
            validatorFunctions.push(requiredValidator);
          }
        }

        if (field === 'text') {
          if (fieldConstraints.minimumLength) {
            const minimumLengthValidator = Validators.makeMinimumLength(
              label,
              fieldConstraints.minimumLength
            );

            validatorFunctions.push(minimumLengthValidator);
          }

          if (fieldConstraints.maximumLength) {
            const maximumLengthValidator = Validators.makeMaximumLength(
              label,
              fieldConstraints.maximumLength
            );

            validatorFunctions.push(maximumLengthValidator);
          }
        }

        if (fieldConstraints.min) {
          const minValueValidator = Validators.makeMinValue(
            label,
            fieldConstraints.min
          );

          validatorFunctions.push(minValueValidator);
        }

        if (fieldConstraints.max) {
          const maxValueValidator = Validators.makeMaxValue(
            label,
            fieldConstraints.max
          );

          validatorFunctions.push(maxValueValidator);
        }

        if (
          field === 'number' &&
          fieldConstraints.fractionLength &&
          fieldConstraints.fractionLength > 0
        ) {
          const patternValidator = Validators.makeNumberFraction(
            label,
            fieldConstraints.fractionLength
          );

          validatorFunctions.push(patternValidator);
        } else if (field === 'number') {
          const patternValidator = Validators.makeNumber(label);

          validatorFunctions.push(patternValidator);
        }
      } else {
        console.warn(
          `@42.nl/jarb-final-form: constraints for "${validator}" not found, but a JarbField was rendered, this should not occur, check your validator.`
        );
      }
    } else {
      console.warn(
        '@42.nl/jarb-final-form: constraints are empty, but a JarbField was rendered, this should not occur, make sure the constraints are loaded before the form is displayed.'
      );
    }

    if (validatorFunctions.length === 0) {
      return null;
    }

    // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
    this.enhancedValidate = async (
      value: FieldValue,
      allValues: object,
      meta?: FieldState<FieldValue>
    ) => {
      // Generate a new async id for this train of async validations.
      this.validationId = Math.random();

      // Store the id locally to this closure.
      const id = this.validationId;

      // Prevent the previous async check from occurring if possible.
      this.debounceResolver(false);

      // Perform synchronous validation
      const results = await Promise.all(
        validatorFunctions.map(validator => validator(value, allValues, meta))
      );

      const errors = results.filter(v => v !== undefined);

      // If there are no synchronous errors, perform the asynchronous validation
      if (errors.length > 0) {
        return errors;
      }

      // We will only perform the asynchronous validation after 200 milliseconds
      // to prevent repeatedly calling the back-end
      const debouncePromise = new Promise<boolean>(resolve => {
        // Prevent the previous async check from occurring if possible.
        this.debounceResolver(false);

        this.debounceResolver = resolve;

        // After a debounce resolve with true.
        window.setTimeout(() => {
          resolve(true);
        }, asyncValidatorsDebounce);
      });

      const shouldPerformAsyncValidation = await debouncePromise;

      if (!shouldPerformAsyncValidation) {
        return undefined;
      }

      const asyncResults = await Promise.all(
        asyncValidatorFunctions.map(validator =>
          validator(value, allValues, meta)
        )
      );

      // Only use the errors when they are still relevant.
      if (this.validationId !== id) {
        return undefined;
      }

      // If there are no errors return undefined to indicate
      // that everything is a-ok.
      const asyncErrors = asyncResults.filter(v => v !== undefined);

      return asyncErrors.length === 0 ? undefined : asyncErrors;
    };

    return this.enhancedValidate;
  }

  public render(): React.ReactNode {
    const validate = this.getEnhancedValidate();

    const props = {
      validate: validate !== null ? validate : undefined,
      ...this.props
    };

    return <Field {...props} />;
  }
}
