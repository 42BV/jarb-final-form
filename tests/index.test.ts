import * as indexBarrel from '../src/index';

test('index barrel', () => {
  expect(indexBarrel).toMatchSnapshot('index-barrel');
});
