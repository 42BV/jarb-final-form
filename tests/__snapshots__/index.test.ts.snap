// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index barrel: index-barrel 1`] = `
{
  "JarbField": [Function],
  "configureConstraint": [Function],
  "getConstraints": [Function],
  "getFieldConstraints": [Function],
  "hasFractionLength": [Function],
  "hasMax": [Function],
  "hasMaximumLength": [Function],
  "hasMin": [Function],
  "hasMinimumLength": [Function],
  "hasPattern": [Function],
  "hasRadix": [Function],
  "isRequired": [Function],
  "loadConstraints": [Function],
  "setConstraints": [Function],
}
`;
